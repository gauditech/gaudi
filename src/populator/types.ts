type PopulatorModel = {
  model: string;
  repeat: { min: number; max: number } | number;
  data: PopulatorData[];
};

type PopulatorData = {
  field: string;
  type: "constant" | "";
};
