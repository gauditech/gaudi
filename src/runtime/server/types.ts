import { NextFunction, Request, Response } from "express";

export type ServerHttpMethods = "all" | "get" | "post" | "put" | "patch" | "delete";

export type ServerRequestHandler =
  | ((request: Request, response: Response) => void | Promise<void>)
  | ((req: Request, resp: Response, next: NextFunction) => void | Promise<void>);

export type EndpointConfig = {
  path: string;
  method: ServerHttpMethods;
  handlers: ServerRequestHandler[];
};

export type ServerMiddlewareNextFn = (err?: unknown) => void;
