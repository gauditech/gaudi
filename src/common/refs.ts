import {
  Definition,
  FieldDef,
  ModelDef,
  QueryDef,
  ReferenceDef,
  RelationDef,
} from "@src/types/definition";

export type RefKind = "model" | "field" | "reference" | "relation" | "query";
export type Ref<T extends RefKind> = T extends "model"
  ? { kind: "model"; value: ModelDef }
  : T extends "field"
  ? { kind: "field"; value: FieldDef }
  : T extends "reference"
  ? { kind: "reference"; value: ReferenceDef }
  : T extends "relation"
  ? { kind: "relation"; value: RelationDef }
  : T extends "query"
  ? { kind: "query"; value: QueryDef }
  : never;

export function getRef<T extends RefKind>(source: Definition | ModelDef[], refKey: string): Ref<T> {
  if ("models" in source) {
    return getRef<T>(source.models, refKey);
  }
  const model = source.find((m) => m.refKey === refKey);
  if (model) return { kind: "model", value: model } as Ref<T>;

  const field = source.flatMap((m) => m.fields).find((f) => f.refKey === refKey);
  if (field) return { kind: "field", value: field } as Ref<T>;

  const reference = source.flatMap((m) => m.references).find((r) => r.refKey === refKey);
  if (reference) return { kind: "reference", value: reference } as Ref<T>;

  const relation = source.flatMap((m) => m.relations).find((r) => r.refKey === refKey);
  if (relation) return { kind: "relation", value: relation } as Ref<T>;

  const query = source.flatMap((m) => m.queries).find((q) => q.refKey === refKey);
  if (query) return { kind: "query", value: query } as Ref<T>;

  throw ["unknown-refkey", refKey];
}

export function getModelProp<T extends RefKind>(model: ModelDef, name: string) {
  return getRef<T>([model], `${model.name}.${name}`);
}

// FIXME first arg should be Definition, not ModelDef[]
export function getTargetModel(models: ModelDef[], refKey: string): ModelDef {
  const prop = getRef(models, refKey);
  switch (prop.kind) {
    case "reference": {
      return getRef<"model">(models, prop.value.toModelRefKey).value;
    }
    case "relation": {
      return getRef<"model">(models, prop.value.fromModelRefKey).value;
    }
    case "query": {
      return getRef<"model">(models, prop.value.retType).value;
    }
    default:
      throw new Error(`Kind ${prop.kind} not supported`);
  }
}

/**
 * This functions maps feild types to target (Prisma) field types
 *
 * https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#model-field-scalar-types
 */
export function getFieldDbType(type: FieldDef["dbtype"]): string {
  switch (type) {
    case "serial":
      return "Int";
    case "integer":
      return "Int";
    case "text":
      return "String";
    case "boolean":
      return "Boolean";
  }
}
