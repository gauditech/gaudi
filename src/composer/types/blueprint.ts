import { Model } from "./model";

export interface Blueprint {
  models: Record<string, Model>;
}
