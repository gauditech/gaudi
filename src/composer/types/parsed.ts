export interface ModelDef {
  name: string;
  fields: FieldDef[];
  references?: ReferenceDef[];
  relations?: RelationDef[];
}

export interface FieldDef {
  name: string;
  type: "string" | "integer" | "datetime";
  nullable?: boolean;
}

export interface ReferenceDef {
  name: string;
  model: string;
}

export interface RelationDef {
  name: string;
  model: string;
}
