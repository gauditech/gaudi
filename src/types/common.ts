export type HookCode = HookInline | HookSource;
export type HookInline = { kind: "inline"; inline: string };
export type HookSource = { kind: "source"; target: string; file: string; runtimeName: string };
