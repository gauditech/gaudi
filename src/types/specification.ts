export type Specification = {
  models: ModelSpec[];
};

type ModelSpec = {
  name: string;
  fields: FieldSpec[];
  references: ReferenceSpec[];
  relations: RelationSpec[];
};

type FieldSpec = {
  name: string;
  type: string;
  unique?: boolean;
  nullable?: boolean;
};

type ReferenceSpec = {
  name: string;
  toModel: string;
  unique?: boolean;
  nullable?: boolean;
};

type RelationSpec = {
  name: string;
  fromModel: string;
  through: string;
};
