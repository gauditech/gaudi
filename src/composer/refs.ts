import { Blueprint } from "./types/blueprint";
import { Field, Model, Reference, Relation } from "./types/model";

export function splitRef(ref: string): [string, string] {
  const path = ref.split(".");
  if (path.length !== 2) {
    throw new Error("Invalid ref");
  }
  return [path[0], path[1]];
}

export function findModel(bp: Blueprint, ref: string): Model {
  return bp.models[ref]!; // TODO: Throw;
}

export function findField(bp: Blueprint, ref: string): Field {
  const [modelName, _fieldName] = splitRef(ref);
  return bp.models[modelName].fields.find((f) => f.selfRef === ref)!; // TODO: Throw;
}

export function findReference(bp: Blueprint, ref: string): Reference {
  const [modelName, _fieldName] = splitRef(ref);
  return bp.models[modelName].references.find((f) => f.selfRef === ref)!; // TODO: Throw;
}

export function findRelation(bp: Blueprint, ref: string): Relation {
  const [modelName, _fieldName] = splitRef(ref);
  return bp.models[modelName].relations.find((f) => f.selfRef === ref)!; // TODO: Throw;
}
