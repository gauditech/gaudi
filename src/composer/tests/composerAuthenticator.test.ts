import { getRef } from "@src/common/refs";
import { compileToOldSpec, compose } from "@src/index";
import { AUTH_TARGET_MODEL_NAME } from "@src/types/specification";

describe("authenticator composer", () => {
  it("succeeds for simple authenticator", () => {
    const bp = `
      auth {
        method basic {}
      }
    `;

    const def = compose(compileToOldSpec(bp));

    expect(def.authenticator).toMatchSnapshot();
    expect(def.models).toMatchSnapshot();
    expect(def.entrypoints).toMatchSnapshot();
  });

  it("resolves authenticator model implicit relations", () => {
    const bp = `
      model UserProfile {
        reference user { to AuthUser }
      }

      auth {
        method basic {}
      }
    `;

    const def = compose(compileToOldSpec(bp));

    const model = getRef.model(def, def.authenticator!.authUserModel.name);
    const relation = model.relations.find((rel) => rel.fromModel === "UserProfile");

    // check authenticator's models
    expect(relation).toMatchSnapshot();
  });
});

describe("authenticator compiler errors", () => {
  it("fails if authenticator model names are already taken", () => {
    const bp1 = `
        model ${AUTH_TARGET_MODEL_NAME} {}

        auth { method basic {} }
      `;

    expect(() => compileToOldSpec(bp1)).toThrowErrorMatchingInlineSnapshot(
      `"Duplicate model definition"`
    );

    const bp2 = `
        model ${AUTH_TARGET_MODEL_NAME}AccessToken {}

        auth { method basic {} }
      `;

    expect(() => compileToOldSpec(bp2)).toThrowErrorMatchingInlineSnapshot(
      `"Duplicate model definition"`
    );
  });

  it("fails if authenticator model names are already taken", () => {
    const bp1 = `
        model ${AUTH_TARGET_MODEL_NAME} {}

        auth { method basic {} }
      `;

    expect(() => compileToOldSpec(bp1)).toThrowErrorMatchingInlineSnapshot(
      `"Duplicate model definition"`
    );

    const bp2 = `
        model ${AUTH_TARGET_MODEL_NAME}AccessToken {}

        auth { method basic {} }
      `;

    expect(() => compileToOldSpec(bp2)).toThrowErrorMatchingInlineSnapshot(
      `"Duplicate model definition"`
    );
  });
});
