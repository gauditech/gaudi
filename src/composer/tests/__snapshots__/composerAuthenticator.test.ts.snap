// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticator composer resolves authenticator model implicit relations 1`] = `
{
  "fromModel": "UserProfile",
  "fromModelRefKey": "UserProfile",
  "kind": "relation",
  "modelRefKey": "AuthUser",
  "name": "userProfileUserRel",
  "refKey": "AuthUser.userProfileUserRel",
  "through": "user",
  "throughRefKey": "UserProfile.user",
  "unique": false,
}
`;

exports[`authenticator composer succeeds for simple authenticator 1`] = `
{
  "accessTokenModel": {
    "name": "AuthUserAccessToken",
    "refKey": "AuthUserAccessToken",
  },
  "authUserModel": {
    "name": "AuthUser",
    "refKey": "AuthUser",
  },
  "method": {
    "kind": "basic",
  },
  "name": "Auth",
}
`;

exports[`authenticator composer succeeds for simple authenticator 2`] = `
[
  {
    "aggregates": [],
    "computeds": [],
    "dbname": "authuser",
    "fields": [
      {
        "dbname": "id",
        "kind": "field",
        "modelRefKey": "AuthUser",
        "name": "id",
        "nullable": false,
        "primary": true,
        "refKey": "AuthUser.id",
        "type": "integer",
        "unique": true,
        "validate": undefined,
      },
      {
        "dbname": "name",
        "kind": "field",
        "modelRefKey": "AuthUser",
        "name": "name",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUser.name",
        "type": "string",
        "unique": false,
        "validate": {
          "args": [
            {
              "kind": "literal",
              "literal": {
                "kind": "integer",
                "value": 1,
              },
            },
          ],
          "kind": "call",
          "validator": "minLength",
        },
      },
      {
        "dbname": "username",
        "kind": "field",
        "modelRefKey": "AuthUser",
        "name": "username",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUser.username",
        "type": "string",
        "unique": true,
        "validate": {
          "args": [
            {
              "kind": "literal",
              "literal": {
                "kind": "integer",
                "value": 8,
              },
            },
          ],
          "kind": "call",
          "validator": "minLength",
        },
      },
      {
        "dbname": "passwordhash",
        "kind": "field",
        "modelRefKey": "AuthUser",
        "name": "passwordHash",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUser.passwordHash",
        "type": "string",
        "unique": false,
        "validate": {
          "args": [
            {
              "kind": "literal",
              "literal": {
                "kind": "integer",
                "value": 8,
              },
            },
          ],
          "kind": "call",
          "validator": "minLength",
        },
      },
    ],
    "hooks": [],
    "kind": "model",
    "name": "AuthUser",
    "queries": [],
    "refKey": "AuthUser",
    "references": [],
    "relations": [
      {
        "fromModel": "AuthUserAccessToken",
        "fromModelRefKey": "AuthUserAccessToken",
        "kind": "relation",
        "modelRefKey": "AuthUser",
        "name": "tokens",
        "refKey": "AuthUser.tokens",
        "through": "authUser",
        "throughRefKey": "AuthUserAccessToken.authUser",
        "unique": false,
      },
    ],
  },
  {
    "aggregates": [],
    "computeds": [],
    "dbname": "authuseraccesstoken",
    "fields": [
      {
        "dbname": "id",
        "kind": "field",
        "modelRefKey": "AuthUserAccessToken",
        "name": "id",
        "nullable": false,
        "primary": true,
        "refKey": "AuthUserAccessToken.id",
        "type": "integer",
        "unique": true,
        "validate": undefined,
      },
      {
        "dbname": "token",
        "kind": "field",
        "modelRefKey": "AuthUserAccessToken",
        "name": "token",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUserAccessToken.token",
        "type": "string",
        "unique": true,
        "validate": undefined,
      },
      {
        "dbname": "expirydate",
        "kind": "field",
        "modelRefKey": "AuthUserAccessToken",
        "name": "expiryDate",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUserAccessToken.expiryDate",
        "type": "string",
        "unique": false,
        "validate": undefined,
      },
      {
        "dbname": "authuser_id",
        "kind": "field",
        "modelRefKey": "AuthUserAccessToken",
        "name": "authUser_id",
        "nullable": false,
        "primary": false,
        "refKey": "AuthUserAccessToken.authUser_id",
        "type": "integer",
        "unique": false,
        "validate": undefined,
      },
    ],
    "hooks": [],
    "kind": "model",
    "name": "AuthUserAccessToken",
    "queries": [],
    "refKey": "AuthUserAccessToken",
    "references": [
      {
        "fieldRefKey": "AuthUserAccessToken.authUser_id",
        "kind": "reference",
        "modelRefKey": "AuthUserAccessToken",
        "name": "authUser",
        "nullable": false,
        "onDelete": undefined,
        "refKey": "AuthUserAccessToken.authUser",
        "toModelFieldRefKey": "AuthUser.id",
        "toModelRefKey": "AuthUser",
        "unique": false,
      },
    ],
    "relations": [],
  },
]
`;

exports[`authenticator composer succeeds for simple authenticator 3`] = `
[
  {
    "endpoints": [
      {
        "actions": [
          {
            "alias": "existingAuthUser",
            "kind": "query",
            "model": "AuthUser",
            "query": {
              "filter": {
                "args": [
                  {
                    "kind": "alias",
                    "namePath": [
                      "AuthUser",
                      "username",
                    ],
                  },
                  {
                    "kind": "variable",
                    "name": "___changeset___username",
                  },
                ],
                "kind": "function",
                "name": "is",
              },
              "fromPath": [
                "AuthUser",
              ],
              "kind": "query",
              "limit": 1,
              "modelRefKey": "AuthUser",
              "name": "$query",
              "offset": undefined,
              "orderBy": undefined,
              "refKey": "N/A",
              "retCardinality": "one",
              "retType": "AuthUser",
              "select": [
                {
                  "alias": "id",
                  "expr": {
                    "kind": "alias",
                    "namePath": [
                      "AuthUser",
                      "id",
                    ],
                  },
                  "kind": "expression",
                  "type": {
                    "kind": "integer",
                    "nullable": false,
                  },
                },
                {
                  "alias": "passwordHash",
                  "expr": {
                    "kind": "alias",
                    "namePath": [
                      "AuthUser",
                      "passwordHash",
                    ],
                  },
                  "kind": "expression",
                  "type": {
                    "kind": "string",
                    "nullable": false,
                  },
                },
              ],
            },
          },
          {
            "alias": "$action_1",
            "hook": {
              "args": [
                {
                  "name": "clearPassword",
                  "setter": {
                    "fieldsetAccess": [
                      "password",
                    ],
                    "kind": "fieldset-input",
                    "required": false,
                    "type": "string",
                  },
                },
                {
                  "name": "hashPassword",
                  "setter": {
                    "kind": "reference-value",
                    "target": {
                      "access": [
                        "passwordHash",
                      ],
                      "alias": "existingAuthUser",
                    },
                  },
                },
              ],
              "hook": {
                "file": "hooks/actions.js",
                "kind": "source",
                "runtimeName": "@GAUDI_INTERNAL",
                "target": "authenticateUser",
              },
            },
            "kind": "execute-hook",
            "responds": false,
          },
          {
            "alias": "accessToken",
            "changeset": [
              {
                "name": "token",
                "setter": {
                  "args": [
                    {
                      "kind": "literal",
                      "literal": {
                        "kind": "integer",
                        "value": 32,
                      },
                    },
                  ],
                  "kind": "function",
                  "name": "cryptoToken",
                },
              },
              {
                "name": "expiryDate",
                "setter": {
                  "args": [
                    {
                      "args": [
                        {
                          "args": [],
                          "kind": "function",
                          "name": "now",
                        },
                        {
                          "kind": "literal",
                          "literal": {
                            "kind": "integer",
                            "value": 3600000,
                          },
                        },
                      ],
                      "kind": "function",
                      "name": "+",
                    },
                  ],
                  "kind": "function",
                  "name": "stringify",
                },
              },
              {
                "name": "authUser_id",
                "setter": {
                  "kind": "reference-value",
                  "target": {
                    "access": [
                      "id",
                    ],
                    "alias": "existingAuthUser",
                  },
                },
              },
            ],
            "isPrimary": false,
            "kind": "create-one",
            "model": "AuthUserAccessToken",
            "select": [
              {
                "alias": "token",
                "expr": {
                  "kind": "alias",
                  "namePath": [
                    "accessToken",
                    "token",
                  ],
                },
                "kind": "expression",
                "type": {
                  "kind": "string",
                  "nullable": false,
                },
              },
              {
                "alias": "id",
                "expr": {
                  "kind": "alias",
                  "namePath": [
                    "accessToken",
                    "id",
                  ],
                },
                "kind": "expression",
                "type": {
                  "kind": "integer",
                  "nullable": false,
                },
              },
            ],
            "targetPath": [
              "AuthUserAccessToken",
            ],
          },
          {
            "alias": "$action_3",
            "hook": {
              "args": [
                {
                  "name": "status",
                  "setter": {
                    "kind": "literal",
                    "literal": {
                      "kind": "integer",
                      "value": 200,
                    },
                  },
                },
                {
                  "name": "body_token",
                  "setter": {
                    "kind": "reference-value",
                    "target": {
                      "access": [
                        "token",
                      ],
                      "alias": "accessToken",
                    },
                  },
                },
              ],
              "hook": {
                "file": "hooks/actions.js",
                "kind": "source",
                "runtimeName": "@GAUDI_INTERNAL",
                "target": "sendResponse",
              },
            },
            "kind": "execute-hook",
            "responds": true,
          },
        ],
        "authSelect": [],
        "authorize": undefined,
        "fieldset": {
          "kind": "record",
          "nullable": false,
          "record": {
            "password": {
              "kind": "field",
              "nullable": false,
              "required": true,
              "type": "string",
              "validate": undefined,
            },
            "username": {
              "kind": "field",
              "nullable": false,
              "required": true,
              "type": "string",
              "validate": undefined,
            },
          },
        },
        "kind": "custom-many",
        "method": "POST",
        "parentContext": [],
        "path": "login",
        "responds": false,
        "response": undefined,
        "target": {
          "alias": "$target_0",
          "kind": "model",
          "name": "AuthUser",
          "namePath": [
            "AuthUser",
          ],
          "retType": "AuthUser",
          "select": [
            {
              "alias": "id",
              "expr": {
                "kind": "alias",
                "namePath": [
                  "AuthUser",
                  "id",
                ],
              },
              "kind": "expression",
              "type": {
                "kind": "integer",
                "nullable": false,
              },
            },
          ],
        },
      },
      {
        "actions": [
          {
            "alias": "accessToken",
            "kind": "query",
            "model": "AuthUserAccessToken",
            "query": {
              "filter": {
                "args": [
                  {
                    "kind": "alias",
                    "namePath": [
                      "AuthUserAccessToken",
                      "token",
                    ],
                  },
                  {
                    "kind": "variable",
                    "name": "___requestAuthToken",
                  },
                ],
                "kind": "function",
                "name": "is",
              },
              "fromPath": [
                "AuthUserAccessToken",
              ],
              "kind": "query",
              "limit": 1,
              "modelRefKey": "AuthUserAccessToken",
              "name": "$query",
              "offset": undefined,
              "orderBy": undefined,
              "refKey": "N/A",
              "retCardinality": "one",
              "retType": "AuthUserAccessToken",
              "select": [
                {
                  "alias": "id",
                  "expr": {
                    "kind": "alias",
                    "namePath": [
                      "AuthUserAccessToken",
                      "id",
                    ],
                  },
                  "kind": "expression",
                  "type": {
                    "kind": "integer",
                    "nullable": false,
                  },
                },
              ],
            },
          },
          {
            "kind": "delete-one",
            "model": "AuthUserAccessToken",
            "targetPath": [
              "accessToken",
            ],
          },
          {
            "alias": "$action_2",
            "hook": {
              "args": [
                {
                  "name": "status",
                  "setter": {
                    "kind": "literal",
                    "literal": {
                      "kind": "integer",
                      "value": 204,
                    },
                  },
                },
              ],
              "hook": {
                "file": "hooks/actions.js",
                "kind": "source",
                "runtimeName": "@GAUDI_INTERNAL",
                "target": "sendResponse",
              },
            },
            "kind": "execute-hook",
            "responds": true,
          },
        ],
        "authSelect": [],
        "authorize": undefined,
        "fieldset": {
          "kind": "record",
          "nullable": false,
          "record": {},
        },
        "kind": "custom-many",
        "method": "POST",
        "parentContext": [],
        "path": "logout",
        "responds": false,
        "response": undefined,
        "target": {
          "alias": "$target_0",
          "kind": "model",
          "name": "AuthUser",
          "namePath": [
            "AuthUser",
          ],
          "retType": "AuthUser",
          "select": [
            {
              "alias": "id",
              "expr": {
                "kind": "alias",
                "namePath": [
                  "AuthUser",
                  "id",
                ],
              },
              "kind": "expression",
              "type": {
                "kind": "integer",
                "nullable": false,
              },
            },
          ],
        },
      },
      {
        "actions": [
          {
            "alias": "authUser",
            "changeset": [
              {
                "name": "name",
                "setter": {
                  "fieldsetAccess": [
                    "authUser",
                    "name",
                  ],
                  "kind": "fieldset-input",
                  "required": true,
                  "type": "string",
                },
              },
              {
                "name": "username",
                "setter": {
                  "fieldsetAccess": [
                    "authUser",
                    "username",
                  ],
                  "kind": "fieldset-input",
                  "required": true,
                  "type": "string",
                },
              },
              {
                "name": "passwordHash",
                "setter": {
                  "args": [
                    {
                      "fieldsetAccess": [
                        "password",
                      ],
                      "kind": "fieldset-input",
                      "required": false,
                      "type": "string",
                    },
                    {
                      "kind": "literal",
                      "literal": {
                        "kind": "integer",
                        "value": 10,
                      },
                    },
                  ],
                  "kind": "function",
                  "name": "cryptoHash",
                },
              },
            ],
            "isPrimary": false,
            "kind": "create-one",
            "model": "AuthUser",
            "select": [
              {
                "alias": "id",
                "expr": {
                  "kind": "alias",
                  "namePath": [
                    "authUser",
                    "id",
                  ],
                },
                "kind": "expression",
                "type": {
                  "kind": "integer",
                  "nullable": false,
                },
              },
              {
                "alias": "name",
                "expr": {
                  "kind": "alias",
                  "namePath": [
                    "authUser",
                    "name",
                  ],
                },
                "kind": "expression",
                "type": {
                  "kind": "string",
                  "nullable": false,
                },
              },
              {
                "alias": "username",
                "expr": {
                  "kind": "alias",
                  "namePath": [
                    "authUser",
                    "username",
                  ],
                },
                "kind": "expression",
                "type": {
                  "kind": "string",
                  "nullable": false,
                },
              },
            ],
            "targetPath": [
              "AuthUser",
            ],
          },
          {
            "alias": "$action_1",
            "hook": {
              "args": [
                {
                  "name": "status",
                  "setter": {
                    "kind": "literal",
                    "literal": {
                      "kind": "integer",
                      "value": 201,
                    },
                  },
                },
                {
                  "name": "body_id",
                  "setter": {
                    "kind": "reference-value",
                    "target": {
                      "access": [
                        "id",
                      ],
                      "alias": "authUser",
                    },
                  },
                },
                {
                  "name": "body_name",
                  "setter": {
                    "kind": "reference-value",
                    "target": {
                      "access": [
                        "name",
                      ],
                      "alias": "authUser",
                    },
                  },
                },
                {
                  "name": "body_username",
                  "setter": {
                    "kind": "reference-value",
                    "target": {
                      "access": [
                        "username",
                      ],
                      "alias": "authUser",
                    },
                  },
                },
              ],
              "hook": {
                "file": "hooks/actions.js",
                "kind": "source",
                "runtimeName": "@GAUDI_INTERNAL",
                "target": "sendResponse",
              },
            },
            "kind": "execute-hook",
            "responds": true,
          },
        ],
        "authSelect": [],
        "authorize": undefined,
        "fieldset": {
          "kind": "record",
          "nullable": false,
          "record": {
            "authUser": {
              "kind": "record",
              "nullable": false,
              "record": {
                "name": {
                  "kind": "field",
                  "nullable": false,
                  "required": true,
                  "type": "string",
                  "validate": {
                    "args": [
                      {
                        "kind": "literal",
                        "literal": {
                          "kind": "integer",
                          "value": 1,
                        },
                      },
                    ],
                    "kind": "call",
                    "validator": "minLength",
                  },
                },
                "username": {
                  "kind": "field",
                  "nullable": false,
                  "required": true,
                  "type": "string",
                  "validate": {
                    "args": [
                      {
                        "kind": "literal",
                        "literal": {
                          "kind": "integer",
                          "value": 8,
                        },
                      },
                    ],
                    "kind": "call",
                    "validator": "minLength",
                  },
                },
              },
            },
            "password": {
              "kind": "field",
              "nullable": false,
              "required": true,
              "type": "string",
              "validate": {
                "args": [
                  {
                    "kind": "literal",
                    "literal": {
                      "kind": "integer",
                      "value": 8,
                    },
                  },
                ],
                "kind": "call",
                "validator": "minLength",
              },
            },
          },
        },
        "kind": "custom-many",
        "method": "POST",
        "parentContext": [],
        "path": "register",
        "responds": false,
        "response": undefined,
        "target": {
          "alias": "$target_0",
          "kind": "model",
          "name": "AuthUser",
          "namePath": [
            "AuthUser",
          ],
          "retType": "AuthUser",
          "select": [
            {
              "alias": "id",
              "expr": {
                "kind": "alias",
                "namePath": [
                  "AuthUser",
                  "id",
                ],
              },
              "kind": "expression",
              "type": {
                "kind": "integer",
                "nullable": false,
              },
            },
          ],
        },
      },
    ],
    "entrypoints": [],
    "name": "",
    "target": {
      "alias": "$target_0",
      "identifyWith": {
        "paramName": "authuser_id",
        "path": [
          "id",
        ],
        "type": "integer",
      },
      "kind": "model",
      "name": "AuthUser",
      "namePath": [
        "AuthUser",
      ],
      "retType": "AuthUser",
    },
  },
]
`;
